### PR TITLE
Update how we calculate correctness score and performance score

### PR DESCRIPTION
**Pull request by jiannanWang**
According to the discussion in #97 , I have updated how we calculate correctness score and performance score

~~Specifically, for `eval_correctness`, the function now returns three values: `return correct == total, correct, total.` This indicates whether the operation passes all tests, along with the number of correct tests and the total number of tests.~~

~~For eval_performance, I introduced a `FAIL_FACTOR` that increases the test time as a penalty when a test fails. Currently, this factor is set to 1.1.~~

The above changes have been reverted. Below are the new commits:

Edit:
For the correctness score calculation, I now use `op_test_data` introduce in #92  to check if all tests pass using `all(data["correctness_score"] for data in op_test_data.values()` to judge if an op is correct. The correctness score is computed as the ratio of correct ops.

For the performance score calculation, the original performance score remains unchanged. I have introduced a new metric: perf_at_p (perf@p for writing), similar to fastp from KernelBench. perf_at_p score is the ratio of ops that are both correct and with a speedup greater than p. For example, if p == 0, perf_at_p score is the same as correctness score. If p == 1, perf_at_p score reflects the ratio of correct ops that are also faster than baseline.

Questions:
@Laurawly Hi Laura! does the perf_at_p score make sense to you? Do you have any suggestions for other performance metrics that would be useful here? Thanks!

### CODE REVIEW

File: BackendBench/eval.py [hunk 1]
```diff
@@ -61,7 +63,7 @@ def eval_correctness(op, impl, tests):
         if eval_correctness_test(op, impl, test):
             correct += 1
         total += 1
-    return correct / total
+    return correct == total, correct, total
```
Comment (PaliC) [inline BackendBench/eval.py:66]: In terms of scoring for correctness correct/total is a pretty natural score (which we can derive from this). I personally would just return correct/total, correct

Imo instead of total I think it would be less error prone to just ensure the suites have a length function to get this information. It doesn't make sense that we have to calculate it manually.

I mostly don't understand why we care about correct == total.

File: BackendBench/eval.py [hunk 2]
```diff
@@ -25,6 +25,8 @@
     exc: {exc}
 """
 
+FAIL_FACTOR = 1.1
```
Comment (PaliC) [inline BackendBench/eval.py:28]: @Laurawly do you think this is useful signal. When it comes to seeing how well kernelfalcon is doing perf wise. I think you're insight is especially useful here.

**Review by PaliC — CHANGES_REQUESTED** [review]
I'm a bit ambivalent about the performance score, but here's my take. Generally, when you are looking at these numbers, you want to incentive people to make them go down. I would either make FAIL_FACTOR bigger (possibly 2.0 or larger) to emphasize correctness or just make them separate by having performance score not include failed tests. 

As an overall score I would recommend using fast_p from kernelbench https://arxiv.org/html/2502.10517v1 with aten as the reference

I am a bit confused on the correctness changes.

File: BackendBench/eval.py (hunk 1 above)
Comment (jiannanWang) [reply BackendBench/eval.py:66]: Whether an op is correct should depend on whether it passes all tests. If one test fails, the op is incorrect. I believe this is also how the kernelbench scores correctness.

Comment (shaahins) [reply BackendBench/eval.py:66]: > Whether an op is correct should depend on whether it passes all tests

agree with this 

correct == total _and_ correct / total and total can be complimentary?

File: BackendBench/eval.py (hunk 2 above)
Comment (jiannanWang) [reply BackendBench/eval.py]: reverted.

File: BackendBench/score.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import torch
+
+
+def fastp(correctness, performance, p=0.8):
+    assert len(correctness) == len(performance), (
+        "correctness and performance must have the same length"
+    )
+    return (
+        torch.where(torch.tensor(correctness).bool(), torch.tensor(performance) > p, 0)
```
[the remark this answers is not in the document]
Comment (jiannanWang) [reply BackendBench/score.py:15]: Updated. The default p is set to 1.0.

I changed the name fastp to perf_at_p (perf@p for writing). I do not use fast_a or fast_aten because I feel the score is a general metrix that can be applied to any performance testing.

**Comment by msaroufim** [comment]
lmk when you're ready for another review

File: BackendBench/score.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
```
Comment (msaroufim) [inline BackendBench/score.py:7]: seems kinda small to have its own file

File: BackendBench/scripts/main.py [hunk 5]
```diff
@@ -243,7 +258,11 @@ def cli(
             results = evaluator.get_results()
 
         for result in results:
-            correctness_score = result.correctness_score
+            correctness_score = all(
+                data["correctness_score"]
+                for data in op_test_data.values()
```
Comment (msaroufim) [inline BackendBench/scripts/main.py:263]: I don't think op_test_data exists in the multiprocess branch

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: BackendBench/score.py (hunk 4 above)
Comment (jiannanWang) [reply BackendBench/score.py]: Moved to eval.py

File: BackendBench/scripts/main.py (hunk 5 above)
Comment (jiannanWang) [reply BackendBench/scripts/main.py]: Thanks! I've fixed by replacing with result.test_data

**Comment by jiannanWang** [comment]
New commits:

1. Move `perf_at_score` to eval.py and the test to test_eval.py
2. fix multiprocessing branch by replacing  `op_test_data` with `result.test_data`.
Run `uv run python BackendBench/scripts/main.py --backend aten --suite opinfo --num-workers 8`
Results:
```
<Logs>
[2025-08-25 15:28:41][INFO][multiprocessing_eval.py] Collected 356 results out of 356 tasks
[2025-08-25 15:28:41][INFO][multiprocessing_eval.py] Shutting down multiprocessing evaluator...
[2025-08-25 15:28:43][INFO][multiprocessing_eval.py] Multiprocessing evaluator shutdown complete
correctness score (mean pass rate over all operators): 0.99
performance score (geomean speedup over all operators): nan
perf@p score (rate of correct samples with a speedup greater than p, p=1.0): 0.00
```

File: BackendBench/scripts/main.py [hunk 6]
```diff
@@ -120,6 +120,12 @@ def setup_logging(log_level):
     type=int,
     help="Number of workers to use for multiprocessing, default to None to disable multiprocessing",
 )
+@click.option(
+    "--p",
+    default=1.0,
+    type=float,
+    help="Performance score threshold for perf@p score calculation",
```
Comment (msaroufim) [inline BackendBench/scripts/main.py:127]: add a comment on whether increasing this number is more or less stringent, it's something that regularly trips people up

File: test/test_eval.py [hunk 7]
```diff
@@ -219,3 +221,42 @@ def __init__(self, args, kwargs):
...
+                baseline_speed = actual_speed * overall_performance
+                fastp_score_orig = fastp_kernel_bench(
+                    overall_correctness, baseline_speed, actual_speed, num_tests, p
+                )
+
+                perf_at_p_score = perf_at_p(
+                    overall_correctness.tolist(), overall_performance.tolist(), p
+                )
+
+                assert torch.allclose(
```
Comment (PaliC) [inline test/test_eval.py:260]: Just add a comment about why the perf@p score is calculated subtly differently, but the test here works. Otherwise, it looks like we can just swap out the two scores.

(something like after averages are calculated it's the same suffices tbh)

File: BackendBench/scripts/main.py [hunk 8]
```diff
@@ -243,7 +257,11 @@ def cli(
             results = evaluator.get_results()
 
         for result in results:
-            correctness_score = result.correctness_score
+            correctness_score = all(
```
Comment (PaliC) [inline BackendBench/scripts/main.py:260]: Can you also add a per@p on the operator level as well in test_data (I'd just do this in eval.py)

**Review by PaliC — APPROVED** [review]
lgtm!  A few minor comments, but thanks for working with us to get the scoring right :)

File: BackendBench/scripts/main.py (hunk 6 above)
Comment (jiannanWang) [reply BackendBench/scripts/main.py:127]: Added.
```
@click.option(
    "--p",
    default=1.0,
    type=float,
    help=(
        "Performance score threshold for perf@p score calculation"
        "Note: Increasing this value makes the threshold more stringent, "
        "requiring a higher speedup to meet the performance criteria."
    )
)
```

File: test/test_eval.py (hunk 7 above)
Comment (jiannanWang) [reply test/test_eval.py:260]: Added.
```
                # Note: The perf@p score calculation here differs subtly from the original fastp score in
                # kernel bench. The original fastp score filters correct samples first, then averages.
                # Here, perf@p averages first, then filters correct samples. Despite this difference,
                # both methods produce equivalent results after averaging, so the test remains valid.
```

**Comment by jiannanWang** [comment]
Thanks! I'm not entirely sure about how perf@k applies at the operator level, but I'm happy to discuss it further and submit a follow-up PR to address that.

File: BackendBench/scripts/main.py (hunk 8 above)
Comment (PaliC) [reply BackendBench/scripts/main.py:264]: ohh kernelbench just has a slew of tests and does fast_p on those. The way you are currently calculating perf@p is that you run it off the aggregated correctness scores / perf scores of ops  rather than all of the individual tests. So really we're calculating things off of the mean of the correctness / perf rather than on the test by test level.

Imo for test_dict we should also calculate it on the test by test level (this was can just call fast_p tbh as its equivalent)

File: BackendBench/scripts/main.py [hunk 9]
```diff
@@ -209,7 +220,14 @@ def cli(
                 test.correctness_tests,
                 test.performance_tests,
             )
-            overall_correctness.append(correctness)
+
+            overall_correctness.append(
```
Comment (PaliC) [inline BackendBench/scripts/main.py:224]: you're calculating correctness score here as an aggregate of all the tests per op. Therefore, we are calculating perf@p on the level of the aggregates per op rather than individual tests as kernelbench does.
